### PR TITLE
OpenAIRE4 oai_openaire metadata format support

### PR DESCRIPTION
**Pull request by paulo-graca**
As decided in our Entities meeting (https://wiki.lyrasis.org/display/DSPACE/2019-11-19+DSpace+7+Entities+WG+Meeting), the XOAI Default Context should only display Publications or non Entity Items. An xoai.xml change was required to address this.

IMPORTANT NOTE: This **PR depends** on https://github.com/DSpace/xoai/pull/68 to fix a transformation exception that is occurring with the current release of XOAI lib. I've already referenced the future release version on this PR (for testing purposes you will need to firstly install that version on your system).

We are proposing to split the XOAI method `retrieveMetadata` in parts to ease comprehension of it and to enable reusing of some parts, but the PR does not demand those changes to be made.
It also proposes an alternate virtual-metadata.xml.openaire file to replace the default one when installers want to use more fields from entities on OAI interface.
This additional virtual metadata will enable to represent something like this in this XML in the oai_openaire metadata format:
```
<datacite:creators>
  <datacite:creator>
    <datacite:creatorName>Evans, R.J.</datacite:creatorName>
    <datacite:affiliation>Institute of Science and Technology</datacite:affiliation>
    <datacite:nameIdentifier nameIdentifierScheme="ORCID"
                    schemeURI="http://orcid.org">
      1234-1234-1234-1234
    </datacite:nameIdentifier>
  </datacite:creator>
</datacite:creators>
```

in order to test it, you must first ensure you have the oai.cfg setting uncommented:
`oai.enabled = true`
(NOTE: you may have to restart the service)

If you want need additional metadata at the oai_openaire metadata format, you need to rename `[/dspace/]config/spring/api/virtual-metadata.xml.openaire` file and replace it with the existing one `[/dspace/]config/spring/api/virtual-metadata.xml` and restart your tomcat service container.

Then you may need to run the OAI import from the command line with the cleaning cache parameter to load all data into your default xoai database/Solr:
`/dspace/bin/dspace oai import -c`

And access the URL:
`http://host/server/oai/openaire4?verb=ListRecords&metadataPrefix=oai_openaire`

EDITED:
This PR still doesn't cover 100% of the guidelines (just the Mandatory and Mandatory when Available elements). It's still missing:

Source - https://openaire-guidelines-for-literature-repository-managers.readthedocs.io/en/v4.0.0/field_source.html
License Condition - https://openaire-guidelines-for-literature-repository-managers.readthedocs.io/en/v4.0.0/field_licensecondition.html
Coverage - https://openaire-guidelines-for-literature-repository-managers.readthedocs.io/en/v4.0.0/field_coverage.html
Geo location - https://openaire-guidelines-for-literature-repository-managers.readthedocs.io/en/v4.0.0/field_geolocation.html
Resource version - https://openaire-guidelines-for-literature-repository-managers.readthedocs.io/en/v4.0.0/field_resourceversion.html
Audience - https://openaire-guidelines-for-literature-repository-managers.readthedocs.io/en/v4.0.0/field_audience.html

That will be addressed later on subsequent PRs.

### CODE REVIEW

**Comment by paulo-graca** [comment]
Just wanted to add the test results of `/dspace/bin/dspace oai import -c` for the current status of this PR:
Without this feature (434 items):
OAI 2.0 manager action ended. It took 36 seconds.

With this feature (434 items):
OAI 2.0 manager action ended. It took 283 seconds.
With this feature 2nd time (434 items):
OAI 2.0 manager action ended. It took 277 seconds.

the result is almost 8 times slower... a different approach (perhaps based on enhancing virtual metadata) will be tried. Every suggestion to pull this is welcome.

**Comment by paulo-graca** [comment]
Commit https://github.com/DSpace/DSpace/pull/2592/commits/019a6c42a33b60b2f0ec54a3f35dc9497d590c4c removed loading relationships for each item in XOAI interface. 
I kept the other code changes in this class because tests resulted in less oai processing time in comparison with the same class in master branch.

**Comment by paulo-graca** [comment]
After the most recent changes, using virtual metadata the processing time decreases to 20sec to process a total of 438 items using:
`dspace/bin/dspace oai import -c`

But an error occurred related with XSL transformation. That error is easily fixed with this PR: https://github.com/DSpace/xoai/pull/68

**Comment by benbosman** [comment]
Hi @paulo-graca , I started reviewing this PR today, but I noticed:
* Travis is currently failing
* I can build this locally, but I don't have a good database with the OpenAIRE relationships to test the functionality
* I really appreciate the cleanup of the ItemUtils to avoid a single method without JavaDocs containing over 200 lines of code. Thanks for splitting this into multiple smaller methods and adding JavaDocs. One question about this: did you also change the functionality in this refactor (anything in particular to verify during the reviewing)
* The changes to the defaultFilter imply e.g. authors are not returned anymore. I'm not sure what the desired default behavior should be. Either way, I would recommend to create a separate filter for OpenAire to ensure the guidelines are still applied when a repository admin wants different behavior for the default filter.

**Comment by paulo-graca** [comment]
> With the database I currently used (which had 1055 items and 1147 relationships) and without the virtual-metadata.xml.openaire4 file (since I didn't have any such relationships), the indexing time increased from 0m51.988s to 1m14.820s when comparing master with this branch. This implies performance is now barely impacted

Weird, that shouldn't occur because no significant changes were done to the indexing process. This solution just uses virtual metadata and proposes an alternative metadata to be used if someone wants more data on the OAI. Did it also occur on second time? Thank you @benbosman for this earlier feedback!

**Comment by paulo-graca** [comment]
> Travis is currently failing
Yes, this is because I've referenced a POM dependency of a specific XOAI release version (3.2.11) that isn't currently available. Perhaps, if you the have time, you can also help with bringing this up - https://github.com/DSpace/xoai/pull/73
Until the release, Travis will continue to fail and you will need to test it manually (as you did). Thank you @benbosman !

**Comment by paulo-graca** [comment]
> I really appreciate the cleanup of the ItemUtils to avoid a single method without JavaDocs containing over 200 lines of code. Thanks for splitting this into multiple smaller methods and adding JavaDocs. One question about this: did you also change the functionality in this refactor (anything in particular to verify during the reviewing)

No, the previous method was simply slipped in parts no additions were made. We initially intend to add a method for adding relationship elements but it was removed in the commit - https://github.com/DSpace/DSpace/commit/019a6c42a33b60b2f0ec54a3f35dc9497d590c4c

Code changes that we made in this PR:
- Debugging message on XOAI.java
- Splitting ItemUtils in readable parts (not really required for this PR to work, can be reverted)
- Adding dspace url field element (to have it available on the XOAI)

Other changes are related with "configurations" like the XSLT on the OAI interface and alternative settings of virtual metadata.
As a enhancement for virtual metadata we can discuss to have for a future version a different configuration based on the "media" type. I can have a specific configuration if content is used on angular, or a different one if it's used on OAI. As an analogy HTML can have use different CSS depending on the context:
```
<link media="screen" href="main.css" />
<link media="print" href="print.css" />
```

**Comment by paulo-graca** [comment]
> The changes to the defaultFilter imply e.g. authors are not returned anymore. I'm not sure what the desired default behavior should be. Either way, I would recommend to create a separate filter for OpenAire to ensure the guidelines are still applied when a repository admin wants different behavior for the default filter.

The solution presented is based on a decision in this meeting - https://wiki.lyrasis.org/display/DSPACE/2019-11-19+DSpace+7+Entities+WG+Meeting 
Repositories that are currently used to expose Publications, if entities are introduced and the default filter if is not changed, the repository will expose mingled Publications, OrgUnits, Persons and external harvesters and aggregators won't know how to deal with this mingled data. So the default behavior should be to only expose Publications on the default context.
An I agree with having a detached filter, specific for OpenAIRE4.
For exposing entities, an alternative and specific context can be addressed. But for that I think it can be addressed in a subsequent PR (I will be glad to propose one) unrelated with OpenAIRE4.

**Comment by tdonohue** [comment]
@paulo-graca : XOAI version 3.2.11 has now been released with this fix https://github.com/DSpace/xoai/pull/68  Please feel free to update the POM to use that release, as it is required by this PR.  (Be aware the 3.2.11 release may take an hour or two to show up in Maven Central, but it'll appear here: http://repo2.maven.org/maven2/org/dspace/xoai/)

**Comment by benbosman** [comment]
Thanks @paulo-graca for your updates:
* I will retest the performance on a test server (I tested it locally, which implies another process may have impacted the performance). I was primarily noticing the performance is still good
* Thanks for specifying the changes to ItemUtils, that's great
* I will be able to test the xsl's and virtual metadata better if you have a database, otherwise it would be hard to perform local tests. This would allow me to finalize my review
* The last commits you made also look good

**Comment by paulo-graca** [comment]
Thank you @benbosman !
I'will be glad to send you my database, but for testing I think it's also useful if you tried without my database, because bugs might occur that way.
You only need to have something like this in your XOAI metadata format to start viewing something useful ([hostname]/server/oai/request?verb=GetRecord&metadataPrefix=xoai&identifier=oai:localhost:123456789/403).
```

    <element name="contributor">
         <element name="author">
            <element name="none">
               <field name="value">Terra, Silvia R.</field>
               <field name="value">Canario, Adelino</field>
               <field name="authority">virtual::226</field>
               <field name="confidence">-1</field>
               <field name="value">Cardoso, João</field>
               <field name="value">Félix, Rute</field>
            </element>
         </element>
```
This PR still doesn't yet cover 100% (I will add this to the description) of the guidelines. I'm still missing:
- Publisher - https://openaire-guidelines-for-literature-repository-managers.readthedocs.io/en/v4.0.0/field_publisher.html
- Source - https://openaire-guidelines-for-literature-repository-managers.readthedocs.io/en/v4.0.0/field_source.html
- Subject - https://openaire-guidelines-for-literature-repository-managers.readthedocs.io/en/v4.0.0/field_subject.html
- License Condition - https://openaire-guidelines-for-literature-repository-managers.readthedocs.io/en/v4.0.0/field_licensecondition.html
- Coverage - https://openaire-guidelines-for-literature-repository-managers.readthedocs.io/en/v4.0.0/field_coverage.html
- Geo location - https://openaire-guidelines-for-literature-repository-managers.readthedocs.io/en/v4.0.0/field_geolocation.html
- Resource version - https://openaire-guidelines-for-literature-repository-managers.readthedocs.io/en/v4.0.0/field_resourceversion.html
- Audience - https://openaire-guidelines-for-literature-repository-managers.readthedocs.io/en/v4.0.0/field_audience.html

Despite of "only have configurations", this PR is becoming large. So I will include in this PR all "configurations" that are Mandatory and "Mandatory when Available" and leave the other transformations for subsequent PRs.

File: dspace/config/crosswalks/oai/metadataFormats/oai_openaire.xsl [hunk 1]
```diff
@@ -0,0 +1,1484 @@
...
+                <xsl:value-of select="false()"/>
+            </xsl:otherwise>
+        </xsl:choose>
+    </xsl:template>
+
+    <!-- 
+        this template will try to look for all "virtual::"
+        and rebuild an "Entity" based on all virtual fields 
+        in the source data.
+        This will retrieve something like
```
Comment (tdonohue) [inline dspace/config/crosswalks/oai/metadataFormats/oai_openaire.xsl]: This final line of the comment seems incomplete? It seems like it is meant to have an example?

File: dspace/config/crosswalks/oai/metadataFormats/oai_openaire.xsl [hunk 2]
```diff
@@ -0,0 +1,1484 @@
...
+            <xsl:when test="$isURL = 'true' and $isHandle = 'false'">
+                <xsl:text>URL</xsl:text>
+            </xsl:when>
+            <xsl:otherwise>
+                <xsl:text>N/A</xsl:text>
+            </xsl:otherwise>
+        </xsl:choose>
+    </xsl:template>
+
+    <xsl:template name="resolveResourceTypeGeneral">
```
Comment (tdonohue) [inline dspace/config/crosswalks/oai/metadataFormats/oai_openaire.xsl]: Please describe this template with a comment. It's *very helpful* to help understand the goal of the template, as this is a large XSL.  I appreciate that you've added such a comment for most templates already.

File: dspace/config/crosswalks/oai/metadataFormats/oai_openaire.xsl [hunk 3]
```diff
@@ -0,0 +1,1484 @@
...
+            <xsl:when test="$lc_dc_type = 'software'">
+                <xsl:text>software</xsl:text>
+            </xsl:when>
+            <xsl:otherwise>
+                other research product
+            </xsl:otherwise>
+        </xsl:choose>
+    </xsl:template>
+
+    <xsl:template name="resolveResourceTypeURI">
```
Comment (tdonohue) [inline dspace/config/crosswalks/oai/metadataFormats/oai_openaire.xsl]: Please describe this template with a comment.

File: dspace/config/crosswalks/oai/metadataFormats/oai_openaire.xsl [hunk 4]
```diff
@@ -0,0 +1,1484 @@
...
+                <xsl:text>http://purl.org/coar/resource_type/c_18wz</xsl:text>
+            </xsl:when>
+            <xsl:when test="$lc_dc_type = 'other type of report'">
+                <xsl:text>http://purl.org/coar/resource_type/c_18wq</xsl:text>
+            </xsl:when>
+            <xsl:when test="$lc_dc_type = 'policy report'">
+                <xsl:text>http://purl.org/coar/resource_type/c_186u</xsl:text>
+            </xsl:when>
+            <xsl:when test="$lc_dc_type = 'project deliverable'">
+
```
Comment (tdonohue) [inline dspace/config/crosswalks/oai/metadataFormats/oai_openaire.xsl]: Please describe this template with a comment.

File: dspace/config/crosswalks/oai/metadataFormats/oai_openaire.xsl [hunk 5]
```diff
@@ -0,0 +1,1484 @@
...
+            <xsl:when test="$lc_dc_date_type='updated'">
+                <xsl:text>Updated</xsl:text>
+            </xsl:when>
+            <xsl:when test="$lc_dc_date_type='valid'">
+                <xsl:text>Valid</xsl:text>
+            </xsl:when>
+        </xsl:choose>
+    </xsl:template>
+
+    <xsl:template name="getRelatedIdentifierType">
```
Comment (tdonohue) [inline dspace/config/crosswalks/oai/metadataFormats/oai_openaire.xsl]: Please describe this template with a comment.

File: dspace/config/crosswalks/oai/metadataFormats/oai_openaire.xsl [hunk 6]
```diff
@@ -0,0 +1,1484 @@
...
+    <xsl:template match="doc:element[@name='bundles']/doc:element[@name='bundle']" mode="oaire">
+       <!-- only consider ORIGINAL bundle -->
+        <xsl:if test="doc:field[@name='name' and text()='ORIGINAL']">
+            <xsl:for-each select="doc:element[@name='bitstreams']/doc:element[@name='bitstream']">
+                <xsl:apply-templates select="." mode="oaire"/>
+            </xsl:for-each>
+        </xsl:if>
+    </xsl:template>
+
+    <xsl:template match="doc:element[@name='bitstreams']/doc:element[@name='bitstream']" mode="oaire">
```
Comment (tdonohue) [inline dspace/config/crosswalks/oai/metadataFormats/oai_openaire.xsl]: Please describe this template with a comment.

File: dspace/config/crosswalks/oai/metadataFormats/oai_openaire.xsl [hunk 7]
```diff
@@ -0,0 +1,1484 @@
...
+    </xsl:template>
+
+    <!-- oaire:fundingReference -->
+    <xsl:template match="doc:element" mode="entity_funding">
+        <oaire:fundingReference>
+            <xsl:apply-templates select="doc:field" mode="entity_funding"/>
+        </oaire:fundingReference>
+    </xsl:template>
+
+    <xsl:template match="doc:field[starts-with(@name,'dc.relation')]" mode="entity_funding">
```
Comment (tdonohue) [inline dspace/config/crosswalks/oai/metadataFormats/oai_openaire.xsl]: There are about 6 templates here in a row that don't have a comment. I really appreciate you commenting what field each template is dealing with.  Could you comment these 6 too?

File: dspace/config/crosswalks/oai/transformers/openaire4.xsl [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+ 
+    <!-- Formatting dc.date.issued -->
+    <xsl:template
+        match="/doc:metadata/doc:element[@name='dc']/doc:element[@name='date']/doc:element[@name='issued']/doc:element/doc:field/text()">
+        <xsl:call-template name="formatdate">
+            <xsl:with-param name="datestr" select="."/>
+        </xsl:call-template>
+    </xsl:template>
+
+    <!-- Modifying and normalizing dc.language -->
```
Comment (tdonohue) [inline dspace/config/crosswalks/oai/transformers/openaire4.xsl]: It might be worth noting in this comment that specifically you are normalizing it into ISO 639-3 (from ISO 639-1).

File: dspace/config/crosswalks/oai/transformers/openaire4.xsl [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+                test="$lc_value = 'metadata only access' or $lc_value = 'closedaccess' or $value = 'http://purl.org/coar/access_right/c_14cb'">
+                <xsl:text>metadata only access</xsl:text>
+            </xsl:when>
+            <xsl:otherwise>
+                <xsl:value-of select="$value"/>
+            </xsl:otherwise>
+        </xsl:choose>
+    </xsl:template>
+
+    <!-- Modifying and normalizing dc.type -->
```
Comment (tdonohue) [inline dspace/config/crosswalks/oai/transformers/openaire4.xsl]: Again, it'd be good to note what "standard" format these are being normalized into.  That'll make it easier to add to or correct later on (as needed)

File: dspace/config/crosswalks/oai/transformers/openaire4.xsl [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+
+    </xsl:template>
+
+
+
+    <!-- AUXILIARY TEMPLATES -->
+
+
+    <!-- Date format -->
+    <xsl:template name="formatdate">
```
Comment (tdonohue) [inline dspace/config/crosswalks/oai/transformers/openaire4.xsl]: What date format is this generating?   Is it YYYY-MM-DD?  We should add this to the comment

File: dspace/config/crosswalks/oai/xoai.xml [hunk 11]
```diff
@@ -70,6 +69,28 @@
...
+
+            - https://openaire-guidelines-for-literature-repository-managers.readthedocs.io/en/v4.0.0/
+
+            There is a limitation over the embargoedEndDate parameter:
+
+            - Predefined DSpace fields don't allow to set this up with a default.
+         -->
+        <Context baseurl="openaire4" name="OpenAIRE 4 Context">
+            <!-- Date format, field prefixes, etc are ensured by the transformer -->
+            <!--Transformer ref="openaire4Transformer"/-->
```
Comment (tdonohue) [inline dspace/config/crosswalks/oai/xoai.xml]: Is there a reason why the transformer is commented out by default?

File: dspace/config/spring/api/virtual-metadata.xml.openaire4 [hunk 12]
```diff
@@ -0,0 +1,448 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (tdonohue) [inline dspace/config/spring/api/virtual-metadata.xml.openaire4]: Why does this file end in an `.openaire4` extension?  Is it to disable this by default?  If so, I wonder if there's an easier way to handle disabling these beans by default.  I'm not sure of how to do this myself, but we could look into perhaps doing it via some form of configuration (as all `dspace.cfg` configs are available in Spring XML configs) if there's a way to do that in Spring.

File: dspace/config/spring/api/virtual-metadata.xml.openaire4 [hunk 13]
```diff
@@ -0,0 +1,448 @@
...
+    <bean class="org.dspace.content.virtual.Collected" id="volumeJournal_title">
+        <property name="fields">
+            <util:list>
+                <value>dc.title</value>
+            </util:list>
+        </property>
+    </bean>
+
+    <util:map id="isVolumeOfJournalMap">
+    </util:map>
```
Comment (tdonohue) [inline dspace/config/spring/api/virtual-metadata.xml.openaire4]: This looks odd, an empty mapping?

**Review by tdonohue — COMMENTED** [review]
@paulo-graca : Overall, this looks good to me. I haven't yet tested it, only did a code review.  While I have a lot of inline comments below, most are just requests for more comments (especially in the metadataFormats XSL, which is large and easier to understand with more comments).  I do have a few other questions as well though.

Overall, this looks good. Just need to give it a test too.

File: dspace/config/spring/api/virtual-metadata.xml.openaire4 (hunk 12 above)
Comment (benbosman) [reply dspace/config/spring/api/virtual-metadata.xml.openaire4]: Although I'd consider it out of scope for this PR, one of the solutions I've considered in the past for creating separate virtual metadata configuration per model would be based on the concept of importing spring configuration
We could perhaps use an empty configuration by default, and support importing the openaire4, or alternative models

**Review by benbosman — COMMENTED** [review]
I've tested this code in much detail with my database and a database from Paulo containing OpenAire data. My findings are:
* Performance of the indexing is indeed not impacted. It's sometimes a bit faster, sometimes a bit slower compared to master
* /server/oai/request?verb=ListRecords&metadataPrefix=oai_dc contains fewer items using the branch due to https://github.com/DSpace/DSpace/pull/2592#issuecomment-566108622, but they can be modified to contain all items without impacting the openaire behavior now
* Using the new branch, the amount of items in /server/oai/request?verb=ListRecords&metadataPrefix=oai_dc is identical to the amount in /server/oai/openaire4?verb=ListRecords&metadataPrefix=oai_openaire, and higher to the amount in /server/oai/openaire?verb=ListRecords&metadataPrefix=oai_dc. I'm not sure whether it's expected to see the openaire4 context contains more items compared to the openaire context
* Using the virtual metadata defined here, I can see a publication containing e.g.
```
      <datacite:creator>
         <datacite:creatorName nameType="Personal">Simmons, Cameron</datacite:creatorName>
         <datacite:familyName>Simmons</datacite:familyName>
         <datacite:givenName>Cameron</datacite:givenName>
         <datacite:nameIdentifier nameIdentifierScheme="DSpace" schemeURI="http://dspace.org">/items/b1b2c768-bda1-448a-a073-fc541e8b24d9</datacite:nameIdentifier>
      </datacite:creator>
```

In conclusion, this looks good, but I still need to review some parts of the code

File: dspace/config/crosswalks/oai/xoai.xml (hunk 11 above)
Comment (paulo-graca) [reply dspace/config/crosswalks/oai/xoai.xml]: Until the transformer be ready there as a reason but currently there is no reason. Thank you @tdonohue for spotting this!

File: dspace/config/spring/api/virtual-metadata.xml.openaire4 (hunk 12 above)
Comment (paulo-graca) [reply dspace/config/spring/api/virtual-metadata.xml.openaire4]: Yes @tdonohue. I'm also not so fun of this solution. And if changing the default virtual metatada configuration file, I may be introducing load of unnecessary data. So the idea here is to have an alternative configuration file, that isn't processed when loading spring beans and can be used (replacing the existing file virtual-metadata.xml) to enhance and to return more openAIRE4 data (otherwise you will have the bare minimum).
As referred in this comment https://github.com/DSpace/DSpace/pull/2592#issuecomment-566099897 an improvement could be done to Virtual Metadata to address the need of having different configurations for each "media" to be used. It could be used a specific configuration in the angular context and a different one for OAI meaning according the the "presentation media" we could use different configurations.

Comment (paulo-graca) [reply dspace/config/spring/api/virtual-metadata.xml.openaire4]: I agree with @benbosman. Changing the virtual metadata could be addressed in a different PR and I'm willing to help on that. And if this PR is merged with this file also change it to better suit the result of that change.

File: dspace/config/spring/api/virtual-metadata.xml.openaire4 (hunk 13 above)
Comment (paulo-graca) [reply dspace/config/spring/api/virtual-metadata.xml.openaire4]: I've just kept what existed in the default virtual-metadata.xml file - https://github.com/DSpace/DSpace/blob/master/dspace/config/spring/api/virtual-metadata.xml#L192

For OpenAIRE4 we don't require Journals or Volumes or Issues and some relations. We can remove all of that. But if this configuration replaces the default file, already existent Entities will loose the display of that data.

File: dspace/config/spring/api/virtual-metadata.xml.openaire4 [hunk 14]
```diff
@@ -0,0 +1,448 @@
...
+                 The keys will define which relationship type will add virtual metadata
+                 to its relationship's item and the map will define which virtual metadata fields
+                 are to be added to the item that has a relationship with this relationship type-->
+            <map>
+                <entry key="isAuthorOfPublication" value-ref="isAuthorOfPublicationMap"/>
+                <entry key="isOrgUnitOfPublication" value-ref="isOrgUnitOfPublicationMap"/>
+                <entry key="isPersonOfProject" value-ref="isPersonOfProjectMap"/>
+                <entry key="isOrgUnitOfProject" value-ref="isOrgUnitOfProjectMap"/>
+                <entry key="isOrgUnitOfPerson" value-ref="isOrgUnitOfPersonMap"/>
+                <entry key="isJournalVolumeOfIssue" value-ref="isJournalVolumeOfIssueMap"/>
```
Comment (benbosman) [inline dspace/config/spring/api/virtual-metadata.xml.openaire4]: Since this is a separate virtual metadata configuration, I would prefer to keep it limited to the openaire-only config

**Review by benbosman — COMMENTED** [review]
I've reviewed the code as well now.
I didn't read through each line of the openaire XSL files, but only checked out how they are built in general. This looks good as well.

I've added one additional inline comment, but other than that, this looks good once the feedback is resolved

File: dspace/config/spring/api/virtual-metadata.xml.openaire4 (hunk 13 above)
Comment (benbosman) [reply dspace/config/spring/api/virtual-metadata.xml.openaire4]: I would prefer to keep this an openaire config file (also mentioned in https://github.com/DSpace/DSpace/pull/2592#discussion_r359338146) and remove these parts

File: dspace/config/crosswalks/oai/transformers/openaire4.xsl (hunk 10 above)
Comment (paulo-graca) [reply dspace/config/crosswalks/oai/transformers/openaire4.xsl]: This is based in an already existent template - https://github.com/DSpace/DSpace/blob/master/dspace/config/crosswalks/oai/transformers/driver.xsl#L57
And I think the purpose it's to discard the  `16:53:24.556` from a date and time like  `2019-04-30 16:53:24.556`. I will add that as a comment.

File: dspace/config/crosswalks/oai/metadataFormats/oai_openaire.xsl (hunk 1 above)
Comment (paulo-graca) [reply dspace/config/crosswalks/oai/metadataFormats/oai_openaire.xsl]: True! I intend to add an example and I forgot to do it :). Just added one.

File: dspace/config/crosswalks/oai/metadataFormats/oai_openaire.xsl (hunk 2 above)
Comment (paulo-graca) [reply dspace/config/crosswalks/oai/metadataFormats/oai_openaire.xsl]: I added comments for each existing templates. Thank you @tdonohue for the review!

File: dspace/config/crosswalks/oai/metadataFormats/oai_openaire.xsl (hunk 7 above)
Comment (paulo-graca) [reply dspace/config/crosswalks/oai/metadataFormats/oai_openaire.xsl]: Just add one description.
https://github.com/DSpace/DSpace/pull/2592/commits/7814a9a1df632f03d9d97191637e722078bba317

File: dspace/config/crosswalks/oai/metadataFormats/oai_openaire.xsl (hunk 6 above)
Comment (paulo-graca) [reply dspace/config/crosswalks/oai/metadataFormats/oai_openaire.xsl]: Just add one description.
https://github.com/DSpace/DSpace/pull/2592/commits/7814a9a1df632f03d9d97191637e722078bba317

File: dspace/config/crosswalks/oai/metadataFormats/oai_openaire.xsl (hunk 5 above)
Comment (paulo-graca) [reply dspace/config/crosswalks/oai/metadataFormats/oai_openaire.xsl]: Just add one description.
https://github.com/DSpace/DSpace/pull/2592/commits/7814a9a1df632f03d9d97191637e722078bba317

File: dspace/config/crosswalks/oai/metadataFormats/oai_openaire.xsl (hunk 3 above)
Comment (paulo-graca) [reply dspace/config/crosswalks/oai/metadataFormats/oai_openaire.xsl]: Just add one description.
https://github.com/DSpace/DSpace/pull/2592/commits/7814a9a1df632f03d9d97191637e722078bba317

File: dspace/config/crosswalks/oai/metadataFormats/oai_openaire.xsl (hunk 4 above)
Comment (paulo-graca) [reply dspace/config/crosswalks/oai/metadataFormats/oai_openaire.xsl]: Just add one description.
https://github.com/DSpace/DSpace/pull/2592/commits/7814a9a1df632f03d9d97191637e722078bba317

File: dspace/config/spring/api/virtual-metadata.xml.openaire4 (hunk 14 above)
Comment (paulo-graca) [reply dspace/config/spring/api/virtual-metadata.xml.openaire4]: I've just removed the unused parts.

**Review by benbosman — APPROVED** [review]
Thanks for the updates @paulo-graca 
I've reviewed it again, and it looks great

**Review by tdonohue — APPROVED** [review]
👍 This looks like a good first version to me.  The code looks good, and I gave it some light testing and it appears to work.

I see only one (minor) thing to cleanup in the future.  We'll want to find a way to rename the `dspace/config/spring/api/virtual-metadata.xml.openaire4` configuration file, as noted in this [comment thread](https://github.com/DSpace/DSpace/pull/2592#discussion_r359013216).  However, I see no reason to hold this up from inclusion for now, and that can be resolved in a future PR once we determine the best strategy for enabling OpenAIRE4

Thanks again @paulo-graca !
